--- xmp-2.0.5-pre3/lib/Makefile.orig	2001-02-18 04:56:08.000000000 +0100
+++ xmp-2.0.5-pre3/lib/Makefile	2003-07-23 21:34:04.000000000 +0200
@@ -23,8 +23,8 @@
 install::
 ifdef DYNDRV
 	[ -d $(LIB_DIR)/xmp/drivers ] || mkdir -p $(LIB_DIR)/xmp/drivers
-	$(INSTALL) -m644 $(MYSHLIB).$(VERSION) $(LIB_DIR)/xmp
-	$(INSTALL) -m644 drivers/*.so $(LIB_DIR)/xmp/drivers
+	$(INSTALL) -m755 $(MYSHLIB).$(VERSION) $(LIB_DIR)
+	$(INSTALL) -m755 drivers/*.so $(LIB_DIR)/xmp/drivers
 endif
 
 shared: $(MYSHLIB).$(VERSION)
@@ -32,7 +32,7 @@
 $(MYSHLIB).$(VERSION): $(MYLIB)
 	rm -f $(MYSHLIB)* *.o
 	$(AR) xv $(MYLIB)
-	ld -shared -o $(MYSHLIB).$(VERSION) *.o
+	$(LD) -shared -o $(MYSHLIB).$(VERSION) *.o -Wl,-soname=$(MYSHLIB).$(VERSION)
 	rm -f *.o
 	ln -s $(MYSHLIB).$(VERSION) $(MYSHLIB)
 	
@@ -40,6 +40,5 @@
 	@echo Installing libraries in $(LIB_DIR)...
 	@[ -d $(LIB_DIR) ] || mkdir -p $(LIB_DIR)
 	$(INSTALL) -m755 $(MYLIB) $(LIB_DIR)
-	$(STRIP) -g $(LIB_DIR)/$(MYLIB)
 
 
--- xmp-2.0.5-pre3/src/drivers/Makefile.orig	2001-02-24 20:32:38.000000000 +0100
+++ xmp-2.0.5-pre3/src/drivers/Makefile	2003-07-23 21:36:13.000000000 +0200
@@ -1,7 +1,7 @@
 # Extended Module Player src/drivers/Makefile
 # $Id$
 
-XCFLAGS	= -I../include
+XCFLAGS	= -I../include -fPIC
 OBJS	= file.o
 DRV	= file.c oss_mix.c hpux.c bsd.c solaris.c oss_seq.c alsa_mix.c esd.c \
 	  sgi.c os2_dart.c netbsd.c openbsd.c aix.c nas.c arts.c qnx.c win32.c
@@ -30,13 +30,13 @@
 	$(CC) $(CFLAGS) $(XCFLAGS) `artsc-config --cflags` -o $@ arts.c
 
 arts.so: arts.o
-	ld -shared -o $@ $+ -lartsc
+	ld -shared -o $@ $+ -lartsc -lc
 
 esd.so: esd.o
-	ld -shared -o $@ $+ -lesd
+	ld -shared -o $@ $+ -lesd -lc
 	
 nas.so: nas.o
-	ld -shared -o $@ $+ -laudio $(X_LIB) -lXt -lX11
+	ld -shared -o $@ $+ -laudio $(X_LIB) -lXt -lX11 -lc
 	
 include depend
 
--- xmp-2.0.5-pre3/src/loaders/Makefile.orig	2001-02-26 01:11:49.000000000 +0100
+++ xmp-2.0.5-pre3/src/loaders/Makefile	2003-07-23 21:37:10.000000000 +0200
@@ -1,7 +1,7 @@
 # Extended Module Player src/loaders/Makefile
 # $Id$
 
-XCFLAGS	= -I../include -Iinclude
+XCFLAGS	= -I../include -Iinclude -fPIC
 LOADERS	= xm_load.o mod_load.o s3m_load.o stm_load.o 669_load.o far_load.o \
 	  mtm_load.o ptm_load.o okt_load.o alm_load.o amd_load.o rad_load.o \
 	  ult_load.o pru1_load.o pru2_load.o kris_load.o ac1d_load.o \
--- xmp-2.0.5-pre3/src/player/Makefile.orig	2001-01-21 19:38:51.000000000 +0100
+++ xmp-2.0.5-pre3/src/player/Makefile	2003-07-23 21:40:21.000000000 +0200
@@ -1,7 +1,7 @@
 # Extended Module Player src/player/Makefile
 # $Id$
 
-XCFLAGS	= -I../include -DLIBDIR='"$(LIB_DIR)"'
+XCFLAGS	= -I../include -DLIBDIR='"$(LIB_DIR)"' -fPIC
 OBJS	= convert.o driver.o formats.o misc.o period.o player.o \
 	  readrc.o scan.o ulaw.o med_synth.o cruncher.o filter.o \
 	  fmopl.o synth.o
--- xmp-2.0.5-pre3/src/misc/Makefile.orig	2001-01-02 23:49:25.000000000 +0100
+++ xmp-2.0.5-pre3/src/misc/Makefile	2003-07-23 21:37:45.000000000 +0200
@@ -1,6 +1,6 @@
 # Extended Module Player src/misc/Makefile
 
-XCFLAGS	= -I../include
+XCFLAGS	= -I../include -fPIC
 OBJS	= control.o info.o ipc.o load.o depack.o unsqsh.o crc32.o mmcmp.o
 XDEPS	=
 XLIBS	= -L../../lib -lxmp
--- xmp-2.0.5-pre3/src/xmms/Makefile.orig	2001-01-29 00:19:21.000000000 +0100
+++ xmp-2.0.5-pre3/src/xmms/Makefile	2003-07-23 21:35:57.000000000 +0200
@@ -1,7 +1,7 @@
 # Extended Module Player src/xmms/Makefile
 # $Id$
 
-XCFLAGS	= -I../include -I../main -DVERSION=\"$(VERSION)\" `xmms-config --cflags` -DXMMS_PLUGIN -D_REENTRANT #-DDEBUG
+XCFLAGS	= -I../include -I../main -DVERSION=\"$(VERSION)\" `xmms-config --cflags` -DXMMS_PLUGIN -D_REENTRANT -fPIC #-DDEBUG
 DFILES	= Makefile README xmp-plugin.h $(OBJS:.o=.c)
 CFILES	= xmp-plugin.so
 INSTDIR	= $(DEST_DIR)`xmms-config --input-plugin-dir`
--- xmp-2.0.5-pre3/Makefile.rules.in.orig	2001-02-26 01:39:23.000000000 +0100
+++ xmp-2.0.5-pre3/Makefile.rules.in	2003-07-23 21:35:05.000000000 +0200
@@ -48,7 +48,7 @@
 	$(CC) $(CFLAGS) -o $*.o $<
 
 .o.so:
-	ld -shared -o $*.so $<
+	ld -shared -o $*.so $< -lc
 
 dummy:
 
